EtherKit/Models/TypedData.swift

```diff
@@ -0,0 +1,394 @@
+//
+//  TypedData.swift
+//  EtherKit
+//
+//  Created by Zac Morris on 2018-06-14.
+//
+import BigInt
+import Foundation
+import Marshal
+
+public struct TypeProperty: Unmarshaling {
+  var name: String
+  var type: String
+
+  public init(object: MarshaledObject) throws {
+    name = try object.value(for: "name")
+    type = try object.value(for: "type")
+  }
+
+  public func encodeType() -> String {
+    return "\(type) \(name)"
+  }
+}
+
+public struct Type: ValueType {
+  public var properties: [TypeProperty]
+
+  public static func value(from object: Any) throws -> Type {
+    let propertyArray: [TypeProperty] = try [TypeProperty].value(from: object)
+    return Type(properties: propertyArray)
+  }
+
+  public func getSubTypes() -> [String] {
+    let standardTypes: [String] = ["bool", "uint", "int", "address", "string", "bytes"]
+
+    let subTypes: [String] = properties.reduce([String](), { arr, prop in
+      var subType = prop.type
+      let matches = standardTypes.contains { type in
+        return subType.starts(with: type)
+      }
+      if !matches {
+        if subType.contains("[") {
+          subType = subType.components(separatedBy: "[")[0]
+        }
+        if !arr.contains(subType) {
+          var newArr: [String] = arr
+          newArr.append(subType)
+          return newArr
+        }
+      }
+      return arr
+    })
+    return subTypes
+  }
+
+  public func encodeProperties() -> String {
+    let encodedProps = properties.map { prop in
+      return prop.encodeType()
+    }
+    return "(" + encodedProps.joined(separator: ",") + ")"
+  }
+}
+
+public struct Types {
```

Comment:
I like this idea. I only ended up adding this struct so it could hold the functions so this seems like a good idea. I'll try it out and see if I like it.